### PR TITLE
Adding siphon vitality to combat-trainer.

### PR DESCRIPTION
**Pull request by jayRyan24**
Helps during transitioning. We may need to set a player-defined threshold, but right now it's set to 90. 

Originally thought this was the case, but this is really rocking and makes using blood burst a viable option as a TM spell with auto cleanup.

### CODE REVIEW

**Comment by jayRyan24** [comment]
Note: This needs to be absorbed AFTER my other PR, #2295. Which, by the way, is working swell.

File: combat-trainer.lic [hunk 1]
```diff
@@ -1127,8 +1127,15 @@ class SpellProcess
...
     return if game_state.casting
-    return unless DRStats.empath?
-    return if health > 95 && @wounds.empty?
+    return unless DRStats.empath? || DRStats.necromancer?
+    return if health > 95 && @wounds.empty? && DRStats.empath?
+    return if health > 90 && DRStats.necromancer?
+
+    if DRStats.necromancer? && @necromancer_healing['Siphon Vitality']
+      @necromancer_healing['Siphon Vitality']['prep'] = 'target'
+      prepare_spell(@necromancer_healing['Siphon Vitality'], game_state, true)
```
Comment (jandersson) [inline combat-trainer.lic]: I dont think true/false should be a parameter to prepare_spell. The intention should be captured well enough by modifying the spell data or game_state, e.g. line 1135

File: combat-trainer.lic [hunk 2]
```diff
@@ -1151,10 +1158,10 @@ class SpellProcess
         data = { 'abbrev' => 'heal', 'mana' => @empath_spells['HEAL'].first, 'cambrinth' => @empath_spells['HEAL'][1..-1] }
       end
       @wounds = {}
-      prepare_spell(data, game_state) if data
+      prepare_spell(data, game_state, true) if data
```
Comment (jandersson) [inline combat-trainer.lic]: See above comment

File: combat-trainer.lic [hunk 3]
```diff
@@ -1151,10 +1158,10 @@ class SpellProcess
...
     elsif @empath_spells['VH']
       data = { 'abbrev' => 'vh', 'mana' => @empath_spells['VH'].first, 'cambrinth' => @empath_spells['VH'][1..-1] }
-      prepare_spell(data, game_state)
+      prepare_spell(data, game_state, true)
```
Comment (jandersson) [inline combat-trainer.lic]: See above comment

File: combat-trainer.lic (hunk 1 above)
Comment (jayRyan24) [reply combat-trainer.lic]: Not sure what the issue is. prepare_spell uses cambrinth or attunement based on a setting and then based on the mind state of said skill, this is a forced cambrinth use for that. I am seriously so confused as to why this is an issue? It makes perfect sense, and defaults to false if not. There is actually no requirement for it to be put through, but if put through, it forces it. I would need a whole other setting for this. That is, unless we change all spells to be able to force cambrinth use regardless of the mind state of attunement vs arcana.

**Comment by rpherbig** [comment]
This feels like it belongs with Consume Flesh and Devour logic (i.e. necromancer healing). Why was this added to `check_health` (previously empath-only) instead?

**Comment by jayRyan24** [comment]
@rpherbig It was added there because it doesn't require a consume ritual being done, the other reason was that the check_health felt like a good place to put a restore of vit that didn't require a ritual. I think check_health could have been there if it didn't have other requirements.

**Comment by jayRyan24** [comment]
What do you think about this instead?

```
  def check_consume(game_state)
    return unless DRStats.necromancer?
    return unless @necromancer_healing
    return if game_state.casting

    if health > @siphon_vit_threshold && @necromancer_healing['Siphon Vitality']
      @necromancer_healing['Siphon Vitality']['prep'] = 'target'
      prepare_spell(@necromancer_healing['Siphon Vitality'], game_state, true)
      return
    end

    return unless game_state.prepare_consume

    echo "@necromancer_healing.key?('Consume Flesh'): #{@necromancer_healing.key?('Consume Flesh')}" if $debug_mode_ct
    echo "@necromancer_healing.key?('Devour'): #{@necromancer_healing.key?('Devour')}" if $debug_mode_ct
    data = @necromancer_healing['Devour'] || @necromancer_healing['Consume Flesh']
    game_state.casting_consume = true
    game_state.prepare_consume = false
    prepare_spell(data, game_state)
  end
```

**Comment by jayRyan24** [comment]
This is tested with a BLB dance script. It dances with BLB, which is maintaining BLB and SV and devour. This is working real well.

https://pastebin.com/uz9qNsY8

**Comment by jayRyan24** [comment]
On quick fix.

File: combat-trainer.lic [hunk 4]
```diff
@@ -1181,8 +1184,14 @@ class SpellProcess
     return unless DRStats.necromancer?
     return unless @necromancer_healing
     return if game_state.casting
-    return unless game_state.prepare_consume
 
+    if health <= @siphon_vit_threshold && @necromancer_healing['Siphon Vitality'] && DRRoom.npcs.any?
```
Comment (rpherbig) [inline combat-trainer.lic]: Should `DRRoom.npcs` here be `game_state.npcs`? The latter will filter out anything in `settings.ignored_npcs` (which I _think _ would include a zombie or construct).

Comment (jayRyan24) [reply combat-trainer.lic]: Yup, good call. Thanks!

Comment (jayRyan24) [reply combat-trainer.lic]: Done.

File: combat-trainer.lic [hunk 5]
```diff
@@ -1181,8 +1184,14 @@ class SpellProcess
     return unless DRStats.necromancer?
     return unless @necromancer_healing
     return if game_state.casting
-    return unless game_state.prepare_consume
 
+    if health <= @siphon_vit_threshold && @necromancer_healing['Siphon Vitality'] && game_state.npcs?
```
Comment (rpherbig) [inline combat-trainer.lic]: I don't think `game_state.npcs?` will work since the function is named `npcs`. Maybe `game_state.npcs.empty?`

**Pull request merged**